File: daemon/src/services/nativeHistoryImporter.ts
/**
 * NativeHistoryImporter — read a CLI's *at-rest* native store and normalize its
 * terminal-phase turns into our `NormalizedEvent`s (R0.5–R0.9).
 *
 * This is the capability a later phase (P3 toggle) invokes on `tty→json` to
 * backfill the turns a user ran in the raw terminal — turns that never flowed
 * through the JSON `runTurn` normalization path and so are invisible in the JSON
 * UI. In P2 NOTHING calls the importer yet; it is a tested, wired-up capability.
 *
 * Each importer is a NEW at-rest envelope adapter, NOT a reuse of the live
 * stream parser (R0.6): the native store has no daemon-synthesized `user` event
 * and no synthetic `result` line, so the adapter (a) emits `user` events from the
 * native store's own user content and (b) derives usage from the native message.
 *
 * Availability is exposed via a per-CLI registry so P3 can GATE the toggle:
 * claude + opencode ship here; cursor + agy are DEFERRED (absent → blocked).
 */

import type { NormalizedEvent } from "../types.js";
import { claudeHistoryImporter } from "../agent-plugins/claudeImport.js";
import { opencodeHistoryImporter } from "../agent-plugins/opencodeImport.js";

/** Inputs to one import pass. */
export interface NativeImportRequest {
  /** OUR session id — stamped on every emitted `NormalizedEvent`. */
  sessionId: string;
  /** The harness chat/session id that locates the native store. */
  agentChatId: string;
  /** Session cwd (worktree/project path) — resolves the native store location. */
  cwd: string;
  /**
   * The native cursor watermark from the previous import — a per-CLI coordinate
   * (claude line index, opencode message timestamp), STORED SEPARATELY from our
   * `logSeq` (R0.5). Undefined ⇒ import from the beginning of the native store.
   */
  watermark?: string;
}

export interface NativeImportResult {
  /** Normalized events for every native turn PAST the watermark, in order. */
  events: NormalizedEvent[];
  /** The native cursor to persist for the next import (see `native_watermark`). */
  nextWatermark: string;
}

/** Per-CLI at-rest adapter. */
export interface NativeHistoryImporter {
  readonly cli: string;
  import(req: NativeImportRequest): Promise<NativeImportResult>;
}

/**
 * The importer registry — the P3 toggle GATE. A CLI present here can import its
 * terminal-phase history; a CLI absent here blocks the tty→json toggle (R1.6).
 * cursor + agy are DEFERRED and intentionally absent.
 */
const IMPORTERS: Record<string, NativeHistoryImporter> = {
  claude: claudeHistoryImporter,
  opencode: opencodeHistoryImporter,
};

/** Resolve the native-history importer for a CLI, or undefined if none exists. */
export function getNativeHistoryImporter(cli: string): NativeHistoryImporter | undefined {
  return IMPORTERS[cli];
}

/** Whether a CLI can import terminal-phase history (the P3 toggle gate). */
export function hasNativeHistoryImporter(cli: string): boolean {
  return cli in IMPORTERS;
}

File: rust/vst-agents/src/native_history_importer.rs
//! NativeHistoryImporter — read a CLI's *at-rest* native store and normalize its
//! terminal-phase turns into `NormalizedEvent`s (R0.5–R0.9). Ports
//! `services/nativeHistoryImporter.ts`.
//!
//! The importer registry is the P3 toggle GATE: claude + opencode ship here;
//! cursor + agy are DEFERRED and intentionally absent.

use vst_types::NormalizedEvent;

use crate::claude_import::claude_history_importer;
use crate::opencode_import::opencode_history_importer;

/// Inputs to one import pass.
#[derive(Clone, Debug)]
pub struct NativeImportRequest {
    /// OUR session id — stamped on every emitted `NormalizedEvent`.
    pub session_id: String,
    /// The harness chat/session id that locates the native store.
    pub agent_chat_id: String,
    /// Session cwd (worktree/project path) — resolves the native store location.
    pub cwd: String,
    /// The native cursor watermark from the previous import (per-CLI coordinate).
    pub watermark: Option<String>,
}

/// Result of one import pass.
#[derive(Clone, Debug)]
pub struct NativeImportResult {
    /// Normalized events for every native turn PAST the watermark, in order.
    pub events: Vec<NormalizedEvent>,
    /// The native cursor to persist for the next import.
    pub next_watermark: String,
}

/// Per-CLI at-rest adapter.
pub trait NativeHistoryImporter: Send + Sync {
    fn cli(&self) -> &'static str;
    fn import(&self, req: &NativeImportRequest) -> NativeImportResult;
}

/// A concrete importer created from a test seam (an injected store path).
pub fn get_native_history_importer(cli: &str) -> Option<&'static dyn NativeHistoryImporter> {
    match cli {
        "claude" => Some(claude_history_importer()),
        "opencode" => Some(opencode_history_importer()),
        _ => None,
    }
}

/// Whether a CLI can import terminal-phase history (the P3 toggle gate).
pub fn has_native_history_importer(cli: &str) -> bool {
    matches!(cli, "claude" | "opencode")
}

// Re-export the store-path resolvers for the test seams.
pub use crate::claude_import::claude_native_store_path;
pub use crate::opencode_import::opencode_native_store_path;
